File: src/critique/DocumentIntegrity.cpp
#include "critique/DocumentIntegrity.h"

#include <algorithm>
#include <sstream>

namespace slm
{
    namespace
    {
        std::vector<std::string> lines(const std::string &text)
        {
            std::vector<std::string> result;
            std::istringstream stream(text);
            std::string line;
            while (std::getline(stream, line))
            {
                if (!line.empty() && line.back() == '\r')
                {
                    line.pop_back();
                }
                result.push_back(line);
            }
            return result;
        }
    }

    int DocumentIntegrity::controlCharacters(const std::string &text)
    {
        int count = 0;
        for (unsigned char character : text)
        {
            if (character < 0x20 && character != static_cast<unsigned char>('\n'))
            {
                ++count;
            }
        }
        return count;
    }

    std::vector<int> DocumentIntegrity::splitMacros(const std::string &text)
    {
        std::vector<int> places;
        const auto rows = lines(text);
        for (std::size_t index = 0; index < rows.size(); ++index)
        {
            std::string row = rows[index];
            while (!row.empty() && (row.back() == '\r' || row.back() == ' '))
            {
                row.pop_back();
            }
            if (row.size() >= 2 && row.back() == '\\' && row[row.size() - 2] != '\\')
            {
                places.push_back(static_cast<int>(index) + 1);
            }
        }
        return places;
    }

    bool DocumentIntegrity::carriesTitle(const std::string &text)
    {
        for (const std::string &line : lines(text))
        {
            if (line.empty())
            {
                continue;
            }
            return line.rfind("# ", 0) == 0 && line.rfind("## ", 0) != 0 && line.size() > 8;
        }
        return false;
    }

    std::vector<std::string> DocumentIntegrity::appendixReferencesOutsideIt(
        const std::string &text)
    {
        const auto all = lines(text);
        std::size_t opensAt = all.size();
        std::vector<std::pair<std::string, std::size_t>> placed;
        for (std::size_t index = 0; index < all.size(); ++index)
        {
            const std::string &line = all[index];
            if (line.rfind("# ", 0) == 0 && line.rfind("## ", 0) != 0 &&
                opensAt == all.size() && line.find("Ek") != std::string::npos)
            {
                opensAt = index;
            }
            if (line.rfind("## ", 0) == 0)
            {
                const std::size_t dot = line.find('.');
                if (dot != std::string::npos && dot > 3)
                {
                    placed.emplace_back(line.substr(3, dot - 3), index);
                }
            }
        }

        std::vector<std::string> wrong;
        for (const std::string &line : all)
        {
            std::size_t at = line.find("Ek ");
            while (at != std::string::npos)
            {
                std::size_t end = at + 3;
                while (end < line.size() && line[end] >= '0' && line[end] <= '9')
                {
                    ++end;
                }
                if (end > at + 3)
                {
                    const std::string wanted = line.substr(at + 3, end - at - 3);
                    for (const auto &entry : placed)
                    {
                        if (entry.first == wanted && entry.second < opensAt &&
                            std::find(wrong.begin(), wrong.end(), wanted) == wrong.end())
                        {
                            wrong.push_back(wanted);
                        }
                    }
                }
                at = line.find("Ek ", end);
            }
        }
        return wrong;
    }

}

File: tools/checks/document_integrity.py
"""Structural checks on a markdown document that have nothing to do with what
it argues, only with whether the file itself survived editing intact.

A TOOL THAT EDITS THROUGH AN ESCAPING LANGUAGE. A tool that edits the text
through a language with backslash escapes can turn a markup macro into the
control character its escape names, silently and without changing the length
of the line; the rendered output merely loses a symbol, so the damage
survives review. Counting stray control characters costs nothing and turns
that class of accident into a failure.

THE ONE ESCAPE THE COUNT ABOVE CANNOT SEE. A backslash followed by the letter
n names the line break itself, so a macro beginning with that letter does not
become a control character: it becomes a real line break, and the rest of
the macro becomes prose at the start of the next line. What that leaves
behind is always the same shape, a line whose last character is a single
backslash.

A REWRITE THAT REBUILDS THE DOCUMENT FROM ITS BLOCKS can drop the block that
was never numbered, and every check keyed on numbers would still pass; the
title is keyed on the thing that has no number.

A REFERENCE THAT RESOLVES IS NOT THEREBY CORRECT. A rewrite that applies a
mapping twice leaves every reference pointing at some section, just not the
intended one, and an existence check passes. A reference announced as an
appendix must land after the heading that opens the appendix, which is a
stronger requirement where one is available.

Mirrors src/critique/DocumentIntegrity.h/.cpp.
"""

from __future__ import annotations


def _lines(text: str) -> list[str]:
    return [line[:-1] if line.endswith("\r") else line for line in text.split("\n")]


def control_characters(text: str) -> int:
    return sum(1 for ch in text if ord(ch) < 0x20 and ch != "\n")


def split_macros(text: str) -> list[int]:
    places: list[int] = []
    for index, raw in enumerate(_lines(text)):
        row = raw.rstrip("\r ")
        if len(row) >= 2 and row[-1] == "\\" and row[-2] != "\\":
            places.append(index + 1)
    return places


def carries_title(text: str) -> bool:
    for line in _lines(text):
        if not line:
            continue
        return line.startswith("# ") and not line.startswith("## ") and len(line) > 8
    return False


def appendix_references_outside_it(text: str) -> list[str]:
    all_lines = _lines(text)
    opens_at = len(all_lines)
    placed: list[tuple[str, int]] = []
    for index, line in enumerate(all_lines):
        if (
            line.startswith("# ")
            and not line.startswith("## ")
            and opens_at == len(all_lines)
            and "Ek" in line
        ):
            opens_at = index
        if line.startswith("## "):
            dot = line.find(".")
            if dot != -1 and dot > 3:
                placed.append((line[3:dot], index))

    wrong: list[str] = []
    for line in all_lines:
        at = line.find("Ek ")
        while at != -1:
            end = at + 3
            while end < len(line) and line[end].isdigit():
                end += 1
            if end > at + 3:
                wanted = line[at + 3 : end]
                for name, position in placed:
                    if name == wanted and position < opens_at and wanted not in wrong:
                        wrong.append(wanted)
            at = line.find("Ek ", end)
    return wrong
